File: examples/canonical-auth/go/internal/routes/groups.go
package routes

import (
	"net/http"

	"canonical-auth/internal/auth"
	"canonical-auth/internal/domain"

	"github.com/labstack/echo/v4"
)

type GroupsRoutes struct {
	Store *domain.InMemoryStore
}

func (r *GroupsRoutes) List(c echo.Context) error {
	authCtx, ok := c.Get("auth").(*auth.AuthContext)
	if !ok || authCtx.TenantID == nil {
		return echo.NewHTTPError(http.StatusForbidden, "tenant scoped route")
	}

	var tenantGroups []domain.Group
	for _, g := range r.Store.Groups {
		if g.TenantID == *authCtx.TenantID {
			tenantGroups = append(tenantGroups, g)
		}
	}

	if tenantGroups == nil {
		tenantGroups = []domain.Group{}
	}

	return c.JSON(http.StatusOK, tenantGroups)
}

func (r *GroupsRoutes) Create(c echo.Context) error {
	authCtx, ok := c.Get("auth").(*auth.AuthContext)
	if !ok || authCtx.TenantID == nil {
		return echo.NewHTTPError(http.StatusForbidden, "tenant scoped route")
	}

	return c.JSON(http.StatusCreated, map[string]string{"message": "Group created"})
}

func (r *GroupsRoutes) AssignPermission(c echo.Context) error {
	authCtx, ok := c.Get("auth").(*auth.AuthContext)
	if !ok || authCtx.TenantID == nil {
		return echo.NewHTTPError(http.StatusForbidden, "tenant scoped route")
	}

	groupID := c.Param("id")

	var group *domain.Group
	for _, g := range r.Store.Groups {
		if g.ID == groupID && g.TenantID == *authCtx.TenantID {
			group = &g
			break
		}
	}

	if group == nil {
		return echo.NewHTTPError(http.StatusNotFound, "group not found in tenant")
	}

	return c.JSON(http.StatusOK, map[string]string{"message": "Permission assigned"})
}

func (r *GroupsRoutes) AssignUser(c echo.Context) error {
	authCtx, ok := c.Get("auth").(*auth.AuthContext)
	if !ok || authCtx.TenantID == nil {
		return echo.NewHTTPError(http.StatusForbidden, "tenant scoped route")
	}

	groupID := c.Param("id")

	var group *domain.Group
	for _, g := range r.Store.Groups {
		if g.ID == groupID && g.TenantID == *authCtx.TenantID {
			group = &g
			break
		}
	}

	if group == nil {
		return echo.NewHTTPError(http.StatusNotFound, "group not found in tenant")
	}

	return c.JSON(http.StatusOK, map[string]string{"message": "User assigned"})
}

File: examples/canonical-auth/rust/src/routes/groups.rs
use std::sync::Arc;
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use serde_json::json;

use crate::auth::claims::AuthContext;
use crate::domain::store::InMemoryStore;

pub async fn list_groups(
    auth_ctx: AuthContext,
    State(_store): State<Arc<InMemoryStore>>,
) -> impl IntoResponse {
    if !auth_ctx.has_permission("iam:group:read") {
        return StatusCode::FORBIDDEN.into_response();
    }
    if auth_ctx.tenant_id.is_none() {
        return StatusCode::FORBIDDEN.into_response();
    }

    (StatusCode::OK, Json(json!([]))).into_response()
}

pub async fn create_group(
    auth_ctx: AuthContext,
    State(_store): State<Arc<InMemoryStore>>,
) -> impl IntoResponse {
    if !auth_ctx.has_permission("iam:group:create") {
        return StatusCode::FORBIDDEN.into_response();
    }
    if auth_ctx.tenant_id.is_none() {
        return StatusCode::FORBIDDEN.into_response();
    }

    (StatusCode::CREATED, Json(json!({"message": "Group created"}))).into_response()
}

pub async fn assign_permission(
    auth_ctx: AuthContext,
    Path(_id): Path<String>,
    State(_store): State<Arc<InMemoryStore>>,
) -> impl IntoResponse {
    if !auth_ctx.has_permission("iam:group:assign_permission") {
        return StatusCode::FORBIDDEN.into_response();
    }
    if auth_ctx.tenant_id.is_none() {
        return StatusCode::FORBIDDEN.into_response();
    }

    (StatusCode::OK, Json(json!({"message": "Permission assigned"}))).into_response()
}

pub async fn assign_user(
    auth_ctx: AuthContext,
    Path(_id): Path<String>,
    State(_store): State<Arc<InMemoryStore>>,
) -> impl IntoResponse {
    if !auth_ctx.has_permission("iam:group:assign_user") {
        return StatusCode::FORBIDDEN.into_response();
    }
    if auth_ctx.tenant_id.is_none() {
        return StatusCode::FORBIDDEN.into_response();
    }

    (StatusCode::OK, Json(json!({"message": "User assigned"}))).into_response()
}
